### ISSUE TITLE
onSelect is sometimes called twice

### ISSUE THREAD

**Issue by woudsma** [Adyen/adyen-web#1832]
**Describe the bug**
The drop-in `onSelect` function is sometimes called twice, usually when selecting either the Klarna or Credit Card payment method.
What's strange is that this doesn't happen every time, it's seemingly random. This makes it hard to debug, I suspect that it might be caused by a race condition.

**To Reproduce**
1. Add a `console.log` to the `onSelect` function.
2. Click on a payment method.
3. Click on another payment method such as Klarna or Credit Card.
4. Keep changing payment methods until you see the `console.log` firing twice when selecting a payment method.

**Expected behavior**
The `onSelect` function is called once after selecting a payment method.

**Screenshots**
Note that we're sending a `PUT` request to our server in the `onSelect` function, you can see the duplicate request being made in the Chrome DevTools network tab.

https://user-images.githubusercontent.com/6162978/200377596-9303002e-bea6-42d8-a11f-5f479dbbc564.mov

**Desktop**
 - OS: macOS Catalina 10.15.7
 - Browser: Chrome, Firefox, Safari
 - Version: Chrome 107.0.5304.87, Firefox 106.0.5, Safari 15.3

**Additional context**
We only define the `onSelect` function **once** when our `AdyenWebDropin` component loads, FYI.
Our initialization looks something like this:
```js
const { configuration, selectHandler } = useDropInConfiguration(i18n)

React.useEffect(() => {
  const initialiseAdyenCheckout = async () => {
    // Create an instance of AdyenCheckout using the configuration object.
    // Remove Afterpay Touch payment method because it is loaded in a separate component
    const checkout = await AdyenCheckout({
      ...configuration,
      removePaymentMethods: [PaymentGroupCodes.AFTERPAY_TOUCH],
    });

    const integration = checkout
      .create('dropin', {
        onSelect: component => {
          console.log('onSelect', new Date().getTime());
          selectHandler(component, integration);
        },
      })
      .mount(dropinContainer.current);

    // Store the checkout to interact with it when payment methods change.
    setActiveCheckout(checkout);

    // Store the web drop in a global context to interact with it elsewhere.
    setActiveWebDropIn(integration);
  };

  initialiseAdyenCheckout();
}, []);
```

<details>
  <summary>Show configuration object and <code class="notranslate">selectHandler</code></summary>
<br />

```js
const configuration = {
  clientKey,
  locale,
  translations,
  environment: environment === 'prd' ? 'live' : 'test',
  paymentMethodsResponse: webDropInPaymentMethods,
  setStatusAutomatically: false,
  showPayButton: true,
  amount: {
    value: adyenAmount,
    currency: totalAfterGiftCards?.currencyIso,
  },
  paymentMethodsConfiguration: {
    bcmc: {
      hasHolderName: true,
      holderNameRequired: true,
      enableStoreDetails: false,
    },
    card: {
      hasHolderName: true,
      holderNameRequired: true,
      enableStoreDetails: !isAnonymous && adyenUseSavedPayment,
    },
    paypal: {
      merchantID,
      environment: environment === 'prd' ? 'live' : 'test',
      onCancel: (data, dropin) => dropin.setStatus('ready'),
      onClick: () => submitOverlappingForm(hasImplicitTermsAndConditions),
      onInit: (data, actions) => paypalOnInitHandler(actions),
    },
    afterpay_default: {},
    klarna: { countryCode: country },
  },
  onSubmit: (state, dropin) => submitHandler(state.data, dropin),
  onAdditionalDetails: (state, dropin) => additionalDetailsHandler(state.data, dropin),
};
```

```js
const selectHandler = async (component, dropin) => {
  try {
    dropin.setStatus('loading');
    
    await dispatch(
      submitPaymentMethod(component?.props?.type, undefined, { isAdyenWebDropInRequest: true })
    );

    dropin.setStatus('ready');
  } catch (error) {
    if (error instanceof OutOfStockError) {
      history?.push(getUrl(Routes.Cart, { locale }));
    }
  }
};
```

<hr />
</details>

Any ideas on how to resolve this issue will be much appreciated!

**Comment by ribeiroguilherme** [comment]
Hey @woudsma ,

Thanks for the detailed explanation. I managed to reproduce the issue on our side, and it seems to be happen randomly indeed. I created one ticket in our backlog to fix this issue.

In the meantime, maybe you could turn the `selectHandler` into a debounce function? So it would be called only once in such scenario where there are 2 or more 'simultaneous' requests calling it

**Comment by woudsma** [comment]
Hi @ribeiroguilherme, thanks for your reply and suggestion.

A debounced `onSelect` handler function does seem to work. You do have to guess what debounce timeout to use so it isn't a perfect solution, but it does the job for now.

For anyone reading, this is our approach:
```js
const integration = checkout
  .create('dropin', {
    onSelect: debounce(component => selectHandler(component, integration), 100),
  })
  .mount(dropinContainer.current);
```

```js
export const debounce = (callback, wait) => {
  let timeout;

  return function () {
    clearTimeout(timeout);
    timeout = setTimeout(() => callback.apply(this, arguments), wait);
  };
};
```

**Comment by woudsma** [comment]
FYI @ribeiroguilherme we rolled back the debounce implementation because it re-introduced a bug where the PayPal button sometimes wouldn't show (a bug that was fixed after we upgraded to v5.29 https://github.com/Adyen/adyen-web/pull/1837). 

A broken PayPal button causes a drop in CVR, so we can't use debounce as a temporary workaround solution for the duplicate `onSelect` calls unfortunately.

**Comment by ribeiroguilherme** [comment]
@woudsma following up about the PayPal problem: can you confirm that, without the debounce implementation, the PayPal button works as expected? 

Regarding the `onSelect`  issue, I will make sure to prioritize this bug fix internally with the team.
